diff -Naurp linux-amlogic-3.10-c954850/drivers/amlogic/ethernet/aml.dtd linux-amlogic-3.10-c954850-patch/drivers/amlogic/ethernet/aml.dtd
--- linux-amlogic-3.10-c954850/drivers/amlogic/ethernet/aml.dtd	2016-02-27 02:12:06.000000000 +0200
+++ linux-amlogic-3.10-c954850-patch/drivers/amlogic/ethernet/aml.dtd	2016-04-04 12:52:17.000000000 +0300
@@ -5,12 +5,21 @@
 		compatible = "amlogic,meson-eth";
 		dev_name = "meson-eth";
 		status = "okay";
-		ethbaseaddr = <0xfe0c0000>;
 		interruptnum = <40>;
-		savepowermode = <0>;
 		phy_interface = <0>;
+		new_maclogic = <1>;
 		reset_pin_enable = <0>;
-		reset_pin = "GPIOY_15";
+		reset_pin = "GPI0DV_28";
 		reset_delay = <100>;
 	};
+
+	ethmac: ethernet@0xc9410000{
+		compatible = "amlogic,meson8m2-rgmii-dwmac", "snps,dwmac";
+		reg = <0xc9410000 0x10000 
+		0xc1108108 0x4>;
+		interrupts = <0 8 1>;
+		interrupt-names = "macirq";
+		clocks = "clk81";
+		clock-names = "stmmaceth";
+	};
 #endif
