diff -Naurp kodi-17.0-beta7-b9fe8e9/xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp kodi-17.0-beta7-b9fe8e9-patch/xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp
--- kodi-17.0-beta7-b9fe8e9/xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp	2016-12-01 15:34:45.000000000 +0200
+++ kodi-17.0-beta7-b9fe8e9-patch/xbmc/cores/VideoPlayer/DVDCodecs/Video/AMLCodec.cpp	2016-12-01 17:18:39.000000000 +0200
@@ -138,7 +138,7 @@ public:
 class DllLibAmCodec : public DllDynamic, DllLibamCodecInterface
 {
   // libamcodec is static linked into libamplayer.so or libamcodec.so
-  DECLARE_DLL_WRAPPER(DllLibAmCodec, "libamplayer.so")
+  DECLARE_DLL_WRAPPER(DllLibAmCodec, "libamcodec.so")
 
   DEFINE_METHOD1(int, codec_init,               (codec_para_t *p1))
   DEFINE_METHOD1(int, codec_close,              (codec_para_t *p1))
@@ -1507,7 +1507,7 @@ CAMLCodec::CAMLCodec()
   m_dll = new DllLibAmCodec;
   if(!m_dll->Load())
   {
-    CLog::Log(LOGWARNING, "CAMLCodec::CAMLCodec libamplayer.so not found, trying libamcodec.so instead");
+    CLog::Log(LOGWARNING, "CAMLCodec::CAMLCodec libamcodec.so not found, trying libamcodec.so instead");
     m_dll->SetFile("libamcodec.so");
     m_dll->Load();
   }
