diff -Naurp linux-d7e124d/drivers/amlogic/amports/vh264.c linux-d7e124d-patch/drivers/amlogic/amports/vh264.c
--- linux-d7e124d/drivers/amlogic/amports/vh264.c	2016-11-30 16:43:57.000000000 +0200
+++ linux-d7e124d-patch/drivers/amlogic/amports/vh264.c	2016-12-18 09:40:47.000000000 +0200
@@ -244,7 +244,7 @@ static bool check_pts_discontinue;
 static u32 wait_buffer_counter;
 static u32 video_signal_from_vui;
 
-static uint error_recovery_mode;
+static uint error_recovery_mode = 1;
 static uint error_recovery_mode_in = 3;
 static uint error_recovery_mode_use = 3;
 
