diff -Naurp linux-amlogic-3.14-f7fde29/arch/arm64/mm/dma-mapping.c linux-amlogic-3.14-f7fde29-patch/arch/arm64/mm/dma-mapping.c
--- linux-amlogic-3.14-f7fde29/arch/arm64/mm/dma-mapping.c	2016-06-11 22:44:33.000000000 +0300
+++ linux-amlogic-3.14-f7fde29-patch/arch/arm64/mm/dma-mapping.c	2016-06-15 16:25:44.000000000 +0300
@@ -45,7 +45,7 @@ static pgprot_t __get_dma_pgprot(struct
 
 static struct gen_pool *atomic_pool;
 
-#define DEFAULT_DMA_COHERENT_POOL_SIZE  SZ_256K
+#define DEFAULT_DMA_COHERENT_POOL_SIZE  SZ_1M
 static size_t atomic_pool_size = DEFAULT_DMA_COHERENT_POOL_SIZE;
 
 static int __init early_coherent_pool(char *p)
