diff -Naurp emulationstation-76c1538/es-app/src/guis/GuiMenu.cpp emulationstation-76c1538-patch/es-app/src/guis/GuiMenu.cpp
--- emulationstation-76c1538/es-app/src/guis/GuiMenu.cpp	2016-08-24 09:39:19.000000000 +0300
+++ emulationstation-76c1538-patch/es-app/src/guis/GuiMenu.cpp	2016-08-24 08:55:39.000000000 +0300
@@ -209,8 +209,31 @@ GuiMenu::GuiMenu(Window* window) : GuiCo
 			mWindow->pushGui(new GuiDetectDevice(mWindow, false, nullptr));
 	});
 
+	addEntry("RETROARCH", 0x777777FF, true, 
+		[this] {
+						remove("/var/lock/start.games");
+						runSystemCommand("touch /var/lock/start.retro");
+						runSystemCommand("systemctl start retroarch.service");
+						SDL_Event ev;
+						ev.type = SDL_QUIT;
+						SDL_PushEvent(&ev);
+	});
+
+	addEntry("REBOOT", 0x777777FF, true,
+		[this] {
+						runRestartCommand();
+	});
+
+	addEntry("SHUTDOWN", 0x777777FF, true, 
+		[this] {
+						runShutdownCommand();
+	});
+
 	addEntry("QUIT", 0x777777FF, true, 
 		[this] {
+						remove("/var/lock/start.games");
+						runSystemCommand("touch /var/lock/start.kodi");
+						runSystemCommand("systemctl start kodi.service");
 						SDL_Event ev;
 						ev.type = SDL_QUIT;
 						SDL_PushEvent(&ev);
diff -Naurp emulationstation-76c1538/es-app/src/main.cpp emulationstation-76c1538-patch/es-app/src/main.cpp
--- emulationstation-76c1538/es-app/src/main.cpp	2016-05-10 01:45:23.000000000 +0300
+++ emulationstation-76c1538-patch/es-app/src/main.cpp	2016-08-24 08:57:54.000000000 +0300
@@ -280,6 +280,9 @@ int main(int argc, char* argv[])
 		window.pushGui(new GuiMsgBox(&window,
 			errorMsg,
 			"QUIT", [] { 
+				remove("/var/lock/start.games");
+				runSystemCommand("touch /var/lock/start.kodi");
+				runSystemCommand("systemctl start kodi.service");
 				SDL_Event* quit = new SDL_Event();
 				quit->type = SDL_QUIT;
 				SDL_PushEvent(quit);
diff -Naurp emulationstation-76c1538/es-core/src/platform.cpp emulationstation-76c1538-patch/es-core/src/platform.cpp
--- emulationstation-76c1538/es-core/src/platform.cpp	2016-05-10 01:45:23.000000000 +0300
+++ emulationstation-76c1538-patch/es-core/src/platform.cpp	2016-08-24 09:02:24.000000000 +0300
@@ -82,26 +82,12 @@ std::string getHomePath()
 
 int runShutdownCommand()
 {
-#if defined(WIN32)
-	return system("shutdown -s -t 0");
-#elif defined(__linux__)
-	sync();
-	return reboot(RB_POWER_OFF);
-#else
-	return system("sudo shutdown -h now");
-#endif
+	return system("systemctl poweroff");
 }
 
 int runRestartCommand()
 {
-#if defined(WIN32)
-	return system("shutdown -r -t 0");
-#elif defined(__linux__)
-	sync();
-	return reboot(RB_AUTOBOOT);
-#else
-	return system("sudo shutdown -r now");
-#endif
+	return system("systemctl reboot");
 }
 
 int runSystemCommand(const std::string& cmd_utf8)
