diff -Naurp tvheadend-3927788/src/input/mpegts/linuxdvb/linuxdvb_private.h tvheadend-3927788-patch/src/input/mpegts/linuxdvb/linuxdvb_private.h
--- tvheadend-3927788/src/input/mpegts/linuxdvb/linuxdvb_private.h	2016-03-13 19:20:43.000000000 +0200
+++ tvheadend-3927788-patch/src/input/mpegts/linuxdvb/linuxdvb_private.h	2016-03-14 09:10:26.000000000 +0200
@@ -438,6 +438,7 @@ int
 linuxdvb_diseqc_send
   (int fd, uint8_t framing, uint8_t addr, uint8_t cmd, uint8_t len, ...);
 int linuxdvb_diseqc_set_volt ( linuxdvb_satconf_t *ls, int volt );
+int linuxdvb_diseqc_set_volt_invert ( linuxdvb_satconf_t *ls, int volt );
 
 /*
  * Satconf
diff -Naurp tvheadend-3927788/src/input/mpegts/linuxdvb/linuxdvb_satconf.c tvheadend-3927788-patch/src/input/mpegts/linuxdvb/linuxdvb_satconf.c
--- tvheadend-3927788/src/input/mpegts/linuxdvb/linuxdvb_satconf.c	2016-03-13 19:20:43.000000000 +0200
+++ tvheadend-3927788-patch/src/input/mpegts/linuxdvb/linuxdvb_satconf.c	2016-03-14 09:15:57.000000000 +0200
@@ -1575,8 +1575,26 @@ linuxdvb_diseqc_set_volt ( linuxdvb_satc
     ls->ls_last_vol = 0;
     return -1;
   }
-  if (vol >= 0)
-    tvh_safe_usleep(15000);
+  tvh_safe_usleep(15000);
+  ls->ls_last_vol = vol ? (vol < 0 ? 0 : 2) : 1;
+  return 0;
+}
+
+int
+linuxdvb_diseqc_set_volt_invert ( linuxdvb_satconf_t *ls, int vol )
+{
+  /* Already set ? */
+  if (vol >= 0 && ls->ls_last_vol == vol + 1)
+    return 0;
+  /* Set voltage */
+  tvhtrace("diseqc", "set voltage %dV", vol ? (vol < 0 ? 0 : 13) : 18);
+  if (ioctl(linuxdvb_satconf_fe_fd(ls), FE_SET_VOLTAGE,
+            vol ? (vol < 0 ? SEC_VOLTAGE_OFF : SEC_VOLTAGE_13) : SEC_VOLTAGE_18)) {
+    tvherror("diseqc", "failed to set voltage (e=%s)", strerror(errno));
+    ls->ls_last_vol = 0;
+    return -1;
+  }
+  tvh_safe_usleep(15000);
   ls->ls_last_vol = vol ? (vol < 0 ? 0 : 2) : 1;
   return 0;
 }
