diff -Naurp retroarch-df4107b/command.c retroarch-df4107b-patch/command.c
--- retroarch-df4107b/command.c	2016-11-11 10:12:26.000000000 +0200
+++ retroarch-df4107b-patch/command.c	2016-11-11 10:37:25.000000000 +0200
@@ -17,6 +17,7 @@
 #include <stdio.h>
 #include <string.h>
 #include <ctype.h>
+#include <fcntl.h>
 
 #ifdef _WIN32
 #include <direct.h>
@@ -1817,9 +1818,6 @@ bool command_event(enum event_command cm
          break;
       case CMD_EVENT_LOAD_CORE:
          command_event(CMD_EVENT_LOAD_CORE_PERSIST, NULL);
-#ifndef HAVE_DYNAMIC
-         command_event(CMD_EVENT_QUIT, NULL);
-#endif
          break;
       case CMD_EVENT_LOAD_STATE:
          /* Immutable - disallow savestate load when
@@ -1925,6 +1923,9 @@ bool command_event(enum event_command cm
 #endif
          break;
       case CMD_EVENT_QUIT:
+         remove("/var/lock/start.retrom");
+         creat("/var/lock/start.kodi", S_IRUSR|S_IROTH);
+         system("systemctl start kodi.service");
          handle_quit_event();
          break;
       case CMD_EVENT_CHEEVOS_HARDCORE_MODE_TOGGLE:
