diff -Naurp media_build_cc-cc56f08/linux/drivers/media/dvb-core/dmxdev.c media_build_cc-cc56f08-patch/linux/drivers/media/dvb-core/dmxdev.c
--- media_build_cc-cc56f08/linux/drivers/media/dvb-core/dmxdev.c	2016-10-17 16:20:26.000000000 +0300
+++ media_build_cc-cc56f08-patch/linux/drivers/media/dvb-core/dmxdev.c	2016-12-08 20:19:32.000000000 +0200
@@ -1128,6 +1128,17 @@ static int dvb_demux_release(struct inod
 	return ret;
 }
 
+#ifdef CONFIG_COMPAT
+static long dvb_demux_compat_ioctl(struct file *filp,
+			unsigned int cmd, unsigned long args)
+{
+	unsigned long ret;
+	args = (unsigned long)compat_ptr(args);
+	ret = dvb_demux_ioctl(filp, cmd, args);
+	return ret;
+}
+#endif
+
 static const struct file_operations dvb_demux_fops = {
 	.owner = THIS_MODULE,
 	.read = dvb_demux_read,
@@ -1136,6 +1147,9 @@ static const struct file_operations dvb_
 	.release = dvb_demux_release,
 	.poll = dvb_demux_poll,
 	.llseek = default_llseek,
+#ifdef CONFIG_COMPAT
+	.compat_ioctl	= dvb_demux_compat_ioctl,
+#endif
 };
 
 static const struct dvb_device dvbdev_demux = {
@@ -1203,6 +1217,18 @@ static unsigned int dvb_dvr_poll(struct
 	return mask;
 }
 
+#ifdef CONFIG_COMPAT
+static long dvb_dvr_compat_ioctl(struct file *filp,
+			unsigned int cmd, unsigned long args)
+{
+	unsigned long ret;
+
+	args = (unsigned long)compat_ptr(args);
+	ret = dvb_dvr_ioctl(filp, cmd, args);
+	return ret;
+}
+#endif
+
 static const struct file_operations dvb_dvr_fops = {
 	.owner = THIS_MODULE,
 	.read = dvb_dvr_read,
@@ -1212,6 +1238,9 @@ static const struct file_operations dvb_
 	.release = dvb_dvr_release,
 	.poll = dvb_dvr_poll,
 	.llseek = default_llseek,
+#ifdef CONFIG_COMPAT
+	.compat_ioctl	= dvb_dvr_compat_ioctl,
+#endif
 };
 
 static const struct dvb_device dvbdev_dvr = {
diff -Naurp media_build_cc-cc56f08/linux/drivers/media/dvb-core/dvb_frontend.c media_build_cc-cc56f08-patch/linux/drivers/media/dvb-core/dvb_frontend.c
--- media_build_cc-cc56f08/linux/drivers/media/dvb-core/dvb_frontend.c	2016-10-17 16:20:26.000000000 +0300
+++ media_build_cc-cc56f08-patch/linux/drivers/media/dvb-core/dvb_frontend.c	2016-12-08 21:22:38.000000000 +0200
@@ -2611,6 +2611,18 @@ static int dvb_frontend_release(struct i
 	return ret;
 }
 
+#ifdef CONFIG_COMPAT
+static long dvb_frontend_compat_ioctl(struct file *filp,
+			unsigned int cmd, unsigned long args)
+{
+	unsigned long ret;
+
+	args = (unsigned long)compat_ptr(args);
+	ret = dvb_generic_ioctl(filp, cmd, args);
+	return ret;
+}
+#endif
+
 static const struct file_operations dvb_frontend_fops = {
 	.owner		= THIS_MODULE,
 	.unlocked_ioctl	= dvb_generic_ioctl,
@@ -2618,6 +2630,9 @@ static const struct file_operations dvb_
 	.open		= dvb_frontend_open,
 	.release	= dvb_frontend_release,
 	.llseek		= noop_llseek,
+#ifdef CONFIG_COMPAT
+	.compat_ioctl	= dvb_frontend_compat_ioctl,
+#endif
 };
 
 int dvb_frontend_suspend(struct dvb_frontend *fe)
diff -Naurp media_build_cc-cc56f08/v4l/compat.h media_build_cc-cc56f08-patch/v4l/compat.h
--- media_build_cc-cc56f08/v4l/compat.h	2016-10-17 16:20:26.000000000 +0300
+++ media_build_cc-cc56f08-patch/v4l/compat.h	2016-12-08 20:44:48.000000000 +0200
@@ -1447,8 +1447,10 @@ static inline bool ether_addr_equal(cons
 #endif
 
 #ifdef NEED_SMP_MB_AFTER_ATOMIC
+#ifndef smp_mb__after_atomic                                                                                                                                                              
 #define smp_mb__after_atomic smp_mb__after_clear_bit
 #endif
+#endif
 
 #ifdef NEED_DEVM_KMALLOC_ARRAY
 static inline void *devm_kmalloc_array(struct device *dev,
