diff -Naurp retroarch-abe2a14/gfx/drivers_context/drm_ctx.c retroarch-abe2a14-patch/gfx/drivers_context/drm_ctx.c
--- retroarch-abe2a14/gfx/drivers_context/drm_ctx.c	2016-07-02 14:49:05.000000000 +0300
+++ retroarch-abe2a14-patch/gfx/drivers_context/drm_ctx.c	2016-07-02 14:25:10.000000000 +0300
@@ -326,8 +326,13 @@ static void free_drm_resources(gfx_ctx_d
    drm_free();
 
    if (drm->drm)
+   {
       if (g_drm_fd >= 0)
+      {
+         drmDropMaster(g_drm_fd);
          filestream_close(drm->drm);
+      }
+   }
 
    g_gbm_surface      = NULL;
    g_gbm_dev          = NULL;
@@ -423,6 +428,8 @@ nextgpu:
    drm->fb_width    = g_drm_connector->modes[0].hdisplay;
    drm->fb_height   = g_drm_connector->modes[0].vdisplay;
 
+   drmSetMaster(g_drm_fd);
+
    g_gbm_dev        = gbm_create_device(fd);
 
    if (!g_gbm_dev)
