diff -Naurp vdr-plugin-streamdev-329129d/client/Makefile vdr-plugin-streamdev-329129d-patch/client/Makefile
--- vdr-plugin-streamdev-329129d/client/Makefile	2013-04-15 10:00:18.000000000 +0300
+++ vdr-plugin-streamdev-329129d-patch/client/Makefile	2013-04-15 07:55:42.000000000 +0300
@@ -26,7 +26,7 @@ CLIENTOBJS = $(PLUGIN).o \
 	
 ### The main target:
 
-all: $(SOFILE) i18n
+all: $(SOFILE) i18n install-i18n
 
 ### Implicit rules:
 
diff -Naurp vdr-plugin-streamdev-329129d/server/Makefile vdr-plugin-streamdev-329129d-patch/server/Makefile
--- vdr-plugin-streamdev-329129d/server/Makefile	2013-04-15 10:00:18.000000000 +0300
+++ vdr-plugin-streamdev-329129d-patch/server/Makefile	2013-04-15 07:56:01.000000000 +0300
@@ -31,7 +31,7 @@ SERVEROBJS = $(PLUGIN).o \
 	
 ### The main target:
 
-all: $(SOFILE) i18n
+all: $(SOFILE) i18n install-i18n
 
 ### Implicit rules:
 
